File: apps/mobile/ios/Visvine/Components/MessageBubble.swift
import SwiftUI

/// One bubble. Mine on the right in the accent; another person's on the left
/// on `surface`; an agent's on the left on `surfaceMuted`. Radius 18 with a
/// tighter tail corner, like Messages.
struct MessageBubble: View {
    @Environment(ThemeStore.self) private var theme
    let text: String
    let isOwn: Bool
    var time: String?
    var senderName: String?
    var agent = false
    /// A failed answer: muted text on the agent tint.
    var failed = false

    var body: some View {
        let c = theme.colors
        HStack(alignment: .bottom, spacing: VVSpace.x2) {
            if isOwn { Spacer(minLength: 48) }
            VStack(alignment: .leading, spacing: VVSpace.x1) {
                if let senderName, !isOwn {
                    Text(senderName).font(.system(size: VVFontSize.s12, weight: .semibold)).foregroundStyle(c.fgMuted)
                }
                Text(text)
                    .font(.system(size: VVFontSize.s16))
                    .foregroundStyle(isOwn ? .white : (failed ? c.fgMuted : c.fg))
                    .textSelection(.enabled)
                if let time {
                    Text(time).font(.system(size: VVFontSize.s10)).foregroundStyle(isOwn ? Color.white.opacity(0.7) : c.fgSubtle)
                }
            }
            .padding(.horizontal, VVSpace.x3_5)
            .padding(.vertical, VVSpace.x2_5)
            .background(fill(c), in: BubbleShape(tail: isOwn ? .right : .left))
            if !isOwn { Spacer(minLength: 48) }
        }
    }

    private func fill(_ c: DynamicColors) -> Color {
        if isOwn { return c.accent }
        return agent ? c.surfaceMuted : c.surface
    }
}

/// 18pt corners with the tail corner at 6pt.
struct BubbleShape: Shape {
    enum Tail { case left, right }
    let tail: Tail

    func path(in rect: CGRect) -> Path {
        let big: CGFloat = 18
        let small: CGFloat = 6
        let bl = tail == .left ? small : big
        let br = tail == .right ? small : big
        var p = Path()
        p.move(to: CGPoint(x: rect.minX + big, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.maxX - big, y: rect.minY))
        p.addQuadCurve(to: CGPoint(x: rect.maxX, y: rect.minY + big), control: CGPoint(x: rect.maxX, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY - br))
        p.addQuadCurve(to: CGPoint(x: rect.maxX - br, y: rect.maxY), control: CGPoint(x: rect.maxX, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX + bl, y: rect.maxY))
        p.addQuadCurve(to: CGPoint(x: rect.minX, y: rect.maxY - bl), control: CGPoint(x: rect.minX, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX, y: rect.minY + big))
        p.addQuadCurve(to: CGPoint(x: rect.minX + big, y: rect.minY), control: CGPoint(x: rect.minX, y: rect.minY))
        p.closeSubpath()
        return p
    }
}

File: apps/mobile/android/app/src/main/java/com/visvine/mobile/ui/components/MessageBubble.kt
package com.visvine.mobile.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.visvine.mobile.ui.theme.VVFontSize
import com.visvine.mobile.ui.theme.VVSpace
import com.visvine.mobile.ui.theme.VisvineTheme

/**
 * One message in a thread. Mine sits right on the accent in white; another's
 * sits left on `bgPrimary`, or on `bgTertiary` when an agent said it. The
 * corner nearest the sender is the 6dp tail. [leading] draws beside another's
 * bubble — an avatar — and nothing beside mine.
 */
@Composable
fun MessageBubble(
    text: String,
    isOwn: Boolean,
    time: String?,
    modifier: Modifier = Modifier,
    senderName: String? = null,
    agent: Boolean = false,
    muted: Boolean = false,
    leading: (@Composable () -> Unit)? = null,
) {
    val colors = VisvineTheme.colors
    val fill = when {
        isOwn -> colors.accent
        agent || muted -> colors.surfaceMuted
        else -> colors.surface
    }
    val ink = when {
        isOwn -> Color.White
        muted -> colors.fgMuted
        else -> colors.fg
    }

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = if (isOwn) Arrangement.End else Arrangement.Start,
    ) {
        if (!isOwn && leading != null) leading()
        Column(
            modifier = Modifier
                .widthIn(max = 280.dp)
                .padding(start = if (isOwn || leading == null) 0.dp else 8.dp)
                .clip(
                    if (isOwn) RoundedCornerShape(18.dp, 18.dp, 6.dp, 18.dp)
                    else RoundedCornerShape(18.dp, 18.dp, 18.dp, 6.dp),
                )
                .background(fill)
                .padding(VVSpace.x3),
        ) {
            if (!isOwn && !senderName.isNullOrBlank()) {
                Text(senderName, color = colors.fgMuted, fontSize = VVFontSize.s12, fontWeight = FontWeight.SemiBold, modifier = Modifier.padding(bottom = VVSpace.x1))
            }
            Text(text, color = ink, fontSize = VVFontSize.s15)
            if (time != null) {
                Text(
                    time,
                    color = if (isOwn) Color.White.copy(alpha = 0.7f) else colors.fgSubtle,
                    fontSize = VVFontSize.s10,
                    modifier = Modifier.padding(top = VVSpace.x1),
                )
            }
        }
    }
}
